File: Game/src/main/java/net/winternetwork/bedwars/game/scheduler/GameScheduler.java
package net.winternetwork.bedwars.game.scheduler;

import net.winternetwork.bedwars.api.module.Module;
import net.winternetwork.bedwars.api.module.ModulePriority;
import net.winternetwork.bedwars.game.Game;
import net.winternetwork.bedwars.game.modules.Modules;
import org.bukkit.Bukkit;

import java.util.Queue;
import java.util.concurrent.ConcurrentLinkedDeque;

public class GameScheduler {

    private final Modules modules = Modules.getInstance();

    public GameScheduler() {
        Bukkit.getScheduler()
                .runTaskTimer(
                        Game.getGame(),
                        this::schedule,
                        0,
                        20L
                );
    }

    private void schedule() {
        Queue<Module> queue = new ConcurrentLinkedDeque<>(modules.getAll());
        for (int i = 0; i < ModulePriority.values().length; i++) {
            for (Module module : queue) {
                if (module.getPriority().getId() != i) continue;

                module.getScheduler().run();
                queue.remove(module);
            }
        }
    }
}

File: Game/src/main/kotlin/net/winternetwork/bedwars/game/scheduler/GameScheduler.kt
package net.winternetwork.bedwars.game.scheduler

import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.async
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import net.winternetwork.bedwars.game.moduleList
import java.util.concurrent.ConcurrentLinkedQueue

class GameScheduler {

    init {
        GlobalScope.launch {
            async {
                while (true) {
                    delay(1000)
                    schedule()
                }
            }.join()
        }
    }

    private fun schedule() {
        val queue = ConcurrentLinkedQueue(moduleList)

        while (!queue.isEmpty()) {
            val module = queue.poll()
            module.onSecPassed()
        }
    }
}